cmd/snap-confine/ns-support.c

```diff
@@ -306,10 +308,172 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
 	die("cannot find mount entry of the root filesystem inside snap namespace");
 }
 
-void sc_create_or_join_ns_group(struct sc_ns_group *group,
-				struct sc_apparmor *apparmor,
-				const char *base_snap_name,
-				const char *snap_name)
+enum sc_discard_vote {
+	SC_DISCARD_NO = 1,
+	SC_DISCARD_YES = 2,
+};
+
+// The namespace may be stale. To check this we must actually switch into it
+// but then we use up our setns call (the kernel misbehaves if we setns twice).
+// To work around this we'll fork a child and use it to probe. The child will
+// inspect the namespace and send information back via eventfd and then exit
+// unconditionally.
+static int sc_inspect_and_maybe_discard_stale_ns(int mnt_fd,
+						 const char *snap_name,
+						 const char *base_snap_name)
+{
+	char base_snap_rev[PATH_MAX] = { 0 };
+	char fname[PATH_MAX] = { 0 };
+	char mnt_fname[PATH_MAX] = { 0 };
+	dev_t base_snap_dev;
+	int event_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	pid_t parent, child;
+
+	// Read the revision of the base snap by looking at the current symlink.
+	sc_must_snprintf(fname, sizeof fname, "%s/%s/current",
+			 SNAP_MOUNT_DIR, base_snap_name);
+	if (readlink(fname, base_snap_rev, sizeof base_snap_rev) < 0) {
+		die("cannot read revision of base snap %s", fname);
+	}
+	if (base_snap_rev[sizeof base_snap_rev - 1] != '\0') {
+		die("cannot use symbolic link %s - value is too long", fname);
+	}
+	// Find the device that is backing the current revision of the base snap.
+	base_snap_dev = find_base_snap_device(base_snap_name, base_snap_rev);
+
+	// Check if we are running on classic. Do it here because we will always
+	// (seemingly) run on a core system once we are inside a mount namespace.
+	bool is_classic = is_running_on_classic_distribution();
+
+	// Store the PID of this process. This is done instead of calls to
+	// getppid() below because then we can reliably track the PID of the
+	// parent even if the child process is re-parented.
+	parent = getpid();
```

Comment:
In this function we declare parent and child up above (and not explicitly initialized), but elsewhere we use:
```
pid_t parent = getpid();
...
pid_t child = fork();
```

It would be nice if you used this here for style consistency.